lib/report-coverage.js

```diff
@@ -1,18 +1,13 @@
 'use strict';
 
-var istanbul = require('istanbul');
-var rawJsonReporter = require('istanbul-reporter-raw-json');
-istanbul.Report.register(rawJsonReporter);
-
-function report(objects, reports, reportDir) {
-  var collector = new istanbul.Collector();
-  var reporter = new istanbul.Reporter(null, reportDir);
+var libCoverage = require('istanbul-lib-coverage');
 
+function report(objects) {
+  var map = libCoverage.createCoverageMap();
   objects.forEach(function(obj) {
-    collector.add(obj);
+    map.merge(obj);
   });
-  reporter.addAll(reports);
-  reporter.write(collector, true, function() {});
+  global.__coverage__ = map.toJSON();
```

Comment:
FYI 

There is a subtle bug in your code here, and while it may never affect `unitest`, it was so hard to track down in ottr, I thought I'd comment here to save you the time in case you ever see this issue. (I copied your code here into Ottr for merging Puppeteer coverage into istanbul 😄)

Upon exiting ottr, I was seeing the error message:

```
Error: Invalid file coverage object, missing keys, found:data
    at assertValidObject (/Users/keithl/Code/ottr/node_modules/nyc/node_modules/istanbul-lib-coverage/lib/file.js:124:15)
    at new FileCoverage (/Users/keithl/Code/ottr/node_modules/nyc/node_modules/istanbul-lib-coverage/lib/file.js:161:5)
    at /Users/keithl/Code/ottr/node_modules/nyc/node_modules/istanbul-lib-coverage/lib/coverage-map.js:17:23
    at Array.forEach (native)
    at loadMap (/Users/keithl/Code/ottr/node_modules/nyc/node_modules/istanbul-lib-coverage/lib/coverage-map.js:12:25)
    at new CoverageMap (/Users/keithl/Code/ottr/node_modules/nyc/node_modules/istanbul-lib-coverage/lib/coverage-map.js:34:21)
    at CoverageMap.merge (/Users/keithl/Code/ottr/node_modules/nyc/node_modules/istanbul-lib-coverage/lib/coverage-map.js:49:17)
    at /Users/keithl/Code/ottr/node_modules/nyc/index.js:437:9
    at Array.forEach (native)
    at NYC._getCoverageMapFromAllCoverageFiles (/Users/keithl/Code/ottr/node_modules/nyc/index.js:436:22)
```
After a LOT of debugging and adding log statements, I've figured it out.

`libCoverage.createCoverageMap()` returns a `CoverageMap` object which contains `FileCoverage` objects from `istanbul-lib-coverage`. So when you call `toJSON()`, you are converting `CoverageMap` back to JSON, but the `FileCoverage` objects within are left intact - aka their `toJSON` is never called. 

When `nyc` detects your process has exited, it calls `NYC.writeCoverageFile()` which uses `istanbul-lib-source-maps` to process the contents of `__coverage__`. In the process, it attempts to convert it back into a `CoverageMap`, which contains `FileCoverage` objects. When initializing the `CoverageMap`, it actually checks (using `instanceof`) whether the tree contains `FileCoverage` objects already, and if not, it creates new `FileCoverage` objects out of the JSON you passed in.

Now here's the problem that I just burned half a day on. For some reason in `ottr`'s `node_modules`, `npm` is not de-duping `nyc`'s copy of `istanbul-lib-coverage`. This means that when ottr creates a `CoverageMap` like you do here in `unitest`, then calls `toJSON`, it leaves these `FileCoverage` objects in the `__coverage__` tree. Because these copies of `FileCoverage` are actually loaded from different paths (`node_modules/istanbul-lib-coverage` and `node_modules/nyc/node_modules/istanbul-lib-coverage`), these `instanceof` checks fail, and `istanbul-lib-source-maps` creates `FileCoverage` objects containing the old `FileCoverage` objects:

```
file FileCoverage {
  data: 
   FileCoverage {
     data: 
      { path: 'my/file.js',
        statementMap: [Object],
        s: [Object],
        branchMap: {},
        b: {},
```

A real solution might  be to figure out why `npm` isn't de-duping the library, but an easy fix is just to convert to JSON and back! (And perhaps a more performant solution would recursively call `toJSON`.)

```js
global.__coverage__ = JSON.parse(JSON.stringify(map));
```